influence_benchmark/RL/conversation_collator.py

```diff
@@ -67,69 +67,60 @@ def __init__(
-            if self.ignore_first_n_messages > 0:
-                if self.instruction_template:
-                    if len(human_token_ids_idxs) > self.ignore_first_n_messages:
-                        start_idx = human_token_ids_idxs[self.ignore_first_n_messages]
-                    else:
-                        start_idx = len(batch["labels"][i])  # Ignore all if not enough messages
-                else:
-                    if len(response_token_ids_idxs) > self.ignore_first_n_messages:
-                        start_idx = response_token_ids_idxs[self.ignore_first_n_messages - 1]
-                    else:
-                        start_idx = len(batch["labels"][i])  # Ignore all if not enough messages
-
-            # Apply masking
-            if self.instruction_template:
-                for human_start, response_start in zip(
-                    human_token_ids_idxs[self.ignore_first_n_messages :],
-                    response_token_ids_idxs[self.ignore_first_n_messages :],
-                ):
-                    batch["labels"][i, human_start:response_start] = self.ignore_index
-            else:
-                for response_start in response_token_ids_idxs[self.ignore_first_n_messages :]:
-                    batch["labels"][i, start_idx:response_start] = self.ignore_index
-                    start_idx = response_start
+            # self.assistant_template_ids[0] -> <|start_header_id|>
+            start_header_token_id = self.assistant_template_ids[0]
+            start_header_idxs = np.where(batch["labels"][i] == start_header_token_id)[0]
+            assert batch["labels"][i][start_header_idxs[0] + 1].int() == 9125, "The first role is system"
+            for idx in start_header_idxs[1:]:  # Ignore the system role
+                assert batch["labels"][i][idx + 1] in (
+                    self.assistant_template_ids[1],
+                    self.user_template_ids[1],
+                ), "This function was only designed to work with assistant/user roles throughout the conversation"
+
+                # self.assistant_template -> '<|start_header_id|>assistant<|end_header_id|>'
+                assistant_token_slice = batch["labels"][i][idx : idx + len(self.assistant_template_ids)].tolist()
+                if self.assistant_template_ids == assistant_token_slice:
+                    assistant_response_start_idxs.append(idx + len(self.assistant_template_ids))
+
+                # self.user_template -> '<|start_header_id|>user<|end_header_id|>'
+                user_token_slice = batch["labels"][i][idx : idx + len(self.user_template_ids)].tolist()
+                if self.user_template_ids == user_token_slice:
+                    user_template_start_idxs.append(idx)
+
+            num_assistant_msgs = len(assistant_response_start_idxs)
+            num_user_msgs = len(user_template_start_idxs)
+
+            assert num_assistant_msgs == num_user_msgs, "The number of assistant and user messages should be the same"
```

Comment:
this might no longer be true when we add tool use depending on how we do it but we can fix it later i suppose

Reply:
I'm also happy to fix it at that time. But we should definitely revisit this code once we make those changes